### PR TITLE
[FLINK-28745][python] Support DataStream PythonCoProcessOperator and PythonKeyedCoProcessOperator in Thread Mode

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request will support DataStream PythonCoProcessOperator and PythonKeyedCoProcessOperator in Thread Mode*

## Brief change log

  - *Support DataStream PythonCoProcessOperator and PythonKeyedCoProcessOperator in Thread Mode*

## Verifying this change

This change added tests and can be verified as follows:

  - *`test_keyed_co_process` and `test_basic_co_operations` in `test_data_stream.py`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by HuangXingBo** [comment]
@flinkbot run azure

**Comment by HuangXingBo** [comment]
@flinkbot run azure

File: flink-python/pyflink/fn_execution/embedded/operations.py [hunk 1]
```diff
@@ -59,17 +66,60 @@ def __init__(self,
...
-            for element in elements:
-                yield from op.process_element(element)
+        results = self._main_operation.process_element(
+            self._input_data_converter.to_internal(value))
 
-        value = [self._input_data_converter.to_internal(value)]
-        for operation in self._operations:
-            value = _process_element(operation, value)
+        for operation in self._chained_operations:
+            results = self._process_element(operation, results)
```
[the remark this answers is not in the document]
Comment (HuangXingBo) [reply flink-python/pyflink/fn_execution/embedded/operations.py]: Yes. Good catch

File: flink-python/pyflink/fn_execution/embedded/operations.py [hunk 2]
```diff
@@ -38,8 +38,15 @@ def close(self):
 
     def on_timer(self, timestamp):
         for operation in self._operations:
-            for item in operation.on_timer(timestamp):
-                yield self._output_data_converter.to_external(item)
+            results = operation.on_timer(timestamp)
```
[the remark this answers is not in the document]
Comment (HuangXingBo) [reply flink-python/pyflink/fn_execution/embedded/operations.py]: Yes. Good catch.

File: flink-python/pyflink/fn_execution/embedded/operations.py [hunk 3]
```diff
@@ -37,8 +37,28 @@ def close(self):
...
+            results = self._process_elements(results)
+
+        if results:
+            yield from self._output_elements(results)
+
+    def _process_elements(self, elements):
+        def _process_elements_on_operation(op, items):
+            if items:
+                for item in items:
+                    yield from op.process_element(item)
```
Comment (Vancior) [inline flink-python/pyflink/fn_execution/embedded/operations.py:52]: Still need to check if the process_element returns None?

**Review by Vancior — APPROVED** [review]

**Pull request closed without merging**